### PR TITLE
[Transformations][ONNX FE][GroupNormalization-12] Decomposition transformation and ONNX FE update

### PR DESCRIPTION
**Pull request by mitruska**
### Details:
 - Add decomposition transformation for GroupNormalization-12
 - Use GroupNormalization-12 for import of org.openvinotoolkit.GroupNorm in ONNX FE (support by decomposition)

### Tickets:
 - 112310, 101520

### CODE REVIEW

File: src/common/transformations/tests/op_conversions/group_normalization_decomposition_test.cpp [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    {
+        function_ref = gen_model_ref({PartialShape{1, 12, 6, 8}, PartialShape{12}, PartialShape{12}},
+                                     element::f32,
+                                     num_groups,
+                                     1e-3f);
+    }
+}
+
+TEST_F(TransformationTestsF, GroupNormalizationDecompositionF16) {
+    const int64_t num_groups = 4;
```
Comment (t-jankowski) [inline src/common/transformations/tests/op_conversions/group_normalization_decomposition_test.cpp:101]: I think that all arguments of the same meaning and value should be instantiated up front to point expected equality. I know it's not common approach in transformations tests, so I don't insist on such "heavy" change.

File: src/common/transformations/src/transformations/op_conversions/group_normalization_decomposition.cpp [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        auto group_norm_node = std::dynamic_pointer_cast<GroupNormalization>(matcher.get_match_root());
+        if (!group_norm_node || transformation_callback(group_norm_node) || group_norm_node->get_input_size() != 3) {
+            return false;
+        }
+
+        auto data = group_norm_node->input_value(0);
+        auto scale = group_norm_node->input_value(1);
+        auto bias = group_norm_node->input_value(2);
+
+        size_t num_groups =
```
Comment (t-jankowski) [inline src/common/transformations/src/transformations/op_conversions/group_normalization_decomposition.cpp:72]:
```suggestion
        const auto num_groups =
```

**Review by t-jankowski — DISMISSED** [review]

File: src/common/transformations/tests/op_conversions/group_normalization_decomposition_test.cpp [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+#include "gtest/gtest.h"
+#include "openvino/core/model.hpp"
+
+using namespace std;
+using namespace ov;
+using namespace opset12;
+using namespace element;
+
+namespace {
+shared_ptr<Model> gen_model(std::vector<PartialShape> input_shapes,
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/common/transformations/tests/op_conversions/group_normalization_decomposition_test.cpp]: Done

File: src/common/transformations/tests/op_conversions/group_normalization_decomposition_test.cpp [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        function =
+            gen_model({PartialShape{1, 12, 6, 8}, PartialShape{12}, PartialShape{12}}, element::f32, num_groups, 1e-3);
+        manager.register_pass<ov::pass::GroupNormalizationDecomposition>();
+    }
+    {
+        function_ref = gen_model_ref({PartialShape{1, 12, 6, 8}, PartialShape{12}, PartialShape{12}},
+                                     element::f32,
+                                     num_groups,
+                                     1e-3f);
+    }
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/common/transformations/tests/op_conversions/group_normalization_decomposition_test.cpp]: Removed nested scopes.

File: src/common/transformations/tests/op_conversions/group_normalization_decomposition_test.cpp [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+
+TEST_F(TransformationTestsF, GroupNormalizationDecompositionF32) {
+    const int64_t num_groups = 4;
+    {
+        function =
+            gen_model({PartialShape{1, 12, 6, 8}, PartialShape{12}, PartialShape{12}}, element::f32, num_groups, 1e-3);
+        manager.register_pass<ov::pass::GroupNormalizationDecomposition>();
+    }
+    {
+        function_ref = gen_model_ref({PartialShape{1, 12, 6, 8}, PartialShape{12}, PartialShape{12}},
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/common/transformations/tests/op_conversions/group_normalization_decomposition_test.cpp]: Renamed to model.

File: src/common/transformations/tests/op_conversions/group_normalization_decomposition_test.cpp (hunk 1 above)
Comment (mitruska) [reply src/common/transformations/tests/op_conversions/group_normalization_decomposition_test.cpp:99]: Done

File: src/common/transformations/src/transformations/op_conversions/group_normalization_decomposition.cpp (hunk 2 above)
Comment (mitruska) [reply src/common/transformations/src/transformations/op_conversions/group_normalization_decomposition.cpp]: Done

File: src/common/transformations/src/transformations/utils/utils.cpp [hunk 6]
```diff
@@ -340,6 +340,19 @@ bool can_eliminate_eltwise_node(const std::shared_ptr<Node>& eltwise,
     }
     return true;
 }
+
+float cast_eps_to_float(double eps_d) {
+    auto eps_f = static_cast<float>(eps_d);
+    if (eps_d > 0.) {  // zero is fine; negative values have no sense
+        if (std::nextafter(eps_d, 0) < static_cast<double>(std::numeric_limits<float>::min()))
```
Comment (praasz) [inline src/common/transformations/src/transformations/utils/utils.cpp:347]: Add unit test for these corner cases and maybe some standard cast value?

Maybe use `ov::cmp::lt`, it should do cast inside.

```suggestion
        if (ov::cmp::lt(std::nextafter(eps_d, 0.), std::numeric_limits<float>::min()))
```

File: src/common/transformations/include/transformations/utils/utils.hpp [hunk 7]
```diff
@@ -87,6 +87,8 @@ inline std::string get_ie_output_name(const Output<Node>& output) {
     return get_ie_output_name(ov::Output<const Node>(output.get_node(), output.get_index()));
 }
 
+float cast_eps_to_float(double eps_d);
```
Comment (praasz) [inline src/common/transformations/include/transformations/utils/utils.hpp:90]: Add description how this cast work, as it is not just simple casting.

File: src/plugins/template/tests/functional/skip_tests_config.cpp [hunk 8]
```diff
@@ -118,7 +118,9 @@ std::vector<std::string> disabledTestPatterns() {
         R"(.*ReferenceInterpolate_v11.*data_type=f16.*)",
         R"(.*LSTMSequence_With_Hardcoded_Refs.*ReferenceLSTMSequenceTest.*iType=f16.*)",
         // CVS-111443
-        R"(.*eltwiseOpType=Mod_secondaryInputType=PARAMETER_opType=VECTOR_NetType=(f16|f32).*)"
+        R"(.*eltwiseOpType=Mod_secondaryInputType=PARAMETER_opType=VECTOR_NetType=(f16|f32).*)",
+        // Interpreter backend doesn't implement evaluate method for OP Multiply (by GroupNormalizationDecomposition)
+        R"(.*ReferenceGroupNormalization.*_f64*)",
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/plugins/template/tests/functional/skip_tests_config.cpp:123]: The set of types supported by evaluate method is intentionally shortened to not increase binary size,
so this test doesn't seem to be a real reason to extend the types supported by evaluate.

File: src/common/transformations/include/transformations/op_conversions/group_normalization_decomposition.hpp [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+namespace ov {
+namespace pass {
+
+class TRANSFORMATIONS_API GroupNormalizationDecomposition;
+
+}  // namespace pass
+}  // namespace ov
+
+// This transformation expresses GroupNormalization with a sub-graph of OpenVINO operations
+class ov::pass::GroupNormalizationDecomposition : public ov::pass::MatcherPass {
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/common/transformations/include/transformations/op_conversions/group_normalization_decomposition.hpp:19]: Such a nice tool, sure I can do it, but to unblock this PR, I would like to add it separately.

File: src/common/transformations/src/transformations/op_conversions/group_normalization_decomposition.cpp [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+#include "openvino/opsets/opset12.hpp"
+#include "openvino/pass/graph_rewrite.hpp"
+#include "openvino/pass/pattern/matcher.hpp"
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+#include "transformations/utils/utils.hpp"
+
+using namespace std;
+using namespace ov;
+using namespace ov::pass;
+using namespace ov::opset12;
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/common/transformations/src/transformations/op_conversions/group_normalization_decomposition.cpp]: Ok, done

File: src/common/transformations/include/transformations/utils/utils.hpp (hunk 7 above)
Comment (mitruska) [reply src/common/transformations/include/transformations/utils/utils.hpp:101]: Description has been added.

File: src/common/transformations/src/transformations/utils/utils.cpp (hunk 6 above)
Comment (mitruska) [reply src/common/transformations/src/transformations/utils/utils.cpp:347]: The eps values are already tested by mvn transformation tests and group norm decomposition (additional added).
Keeping existing version for readability.

File: src/common/transformations/src/transformations/utils/utils.cpp [hunk 11]
```diff
@@ -340,6 +340,19 @@ bool can_eliminate_eltwise_node(const std::shared_ptr<Node>& eltwise,
     }
     return true;
 }
+
+float cast_eps_to_float(double eps_d) {
+    auto eps_f = static_cast<float>(eps_d);
+    if (eps_d > 0.) {  // zero is fine; negative values have no sense
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/common/transformations/src/transformations/utils/utils.cpp:346]: Added a comment in a function description, this solution is used and approved for mvn decomposition.
Additional assert can break backward compatibility, so I would like to avoid significant changes in this helper.

File: src/common/transformations/src/transformations/op_conversions/group_normalization_decomposition.cpp [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    auto num_groups_const = reg.add(Constant::create(element::i64, Shape{1}, {num_groups}));
+    // The 4D shape: [N * num_groups, C // num_groups, H, W] is created
+    // instead of 5D shape: [N, num_groups, C // num_groups, H, W].
+    // The reason is the lack of support for 5D MVN input by some plugins.
+    OutputVector new_shape{reg.make<Multiply>(splits[0], num_groups_const),
+                           reg.make<Divide>(splits[1], num_groups_const)};
+
+    for (size_t i = 2; i < rank_size; ++i) {
+        new_shape.push_back(splits[i]);
+    }
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/common/transformations/src/transformations/op_conversions/group_normalization_decomposition.cpp]: Done

File: src/frontends/onnx/frontend/src/op/org.openvinotoolkit/group_norm.cpp [hunk 13]
```diff
@@ -57,43 +29,14 @@ OutputVector group_norm(const Node& node) {
...
-
-    const auto data_rank = std::make_shared<default_opset::ShapeOf>(data_shape_node);
-
-    if (scale_rank == 1) {
-        result =
-            std::make_shared<default_opset::Multiply>(result,
-                                                      reshape::reshape_channel_shaped_node_to_nchw(scale, data_rank));
-    } else {
-        result = std::make_shared<default_opset::Multiply>(result, scale);
+    if (scale.get_partial_shape().rank().is_static() && scale.get_partial_shape().size() != 1) {
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/frontends/onnx/frontend/src/op/org.openvinotoolkit/group_norm.cpp]: Updated to use compatible(1).
The expected rank is 1, so intention of this Squeeze is to make 1D from shape like [1, 4, 1, 1] as example. 
It is ok to call Squeeze on other shapes, as the result shape will be validated by GropNormalization shape inference.

File: src/common/transformations/tests/op_conversions/group_normalization_decomposition_test.cpp [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+//
+
+#include "transformations/op_conversions/group_normalization_decomposition.hpp"
+
+#include <gtest/gtest.h>
+
+#include <memory>
+#include <openvino/opsets/opset12.hpp>
+#include <openvino/pass/manager.hpp>
+#include <openvino/pass/visualize_tree.hpp>
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/common/transformations/tests/op_conversions/group_normalization_decomposition_test.cpp]: Done

File: src/common/transformations/tests/op_conversions/group_normalization_decomposition_test.cpp [hunk 15]
```diff
@@ -0,0 +1,214 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "transformations/op_conversions/group_normalization_decomposition.hpp"
+
+#include <gtest/gtest.h>
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/common/transformations/tests/op_conversions/group_normalization_decomposition_test.cpp:7]: Done

File: src/common/transformations/tests/op_conversions/group_normalization_decomposition_test.cpp [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+
+TEST_F(TransformationTestsF, GroupNormalizationDecomposition_data_rank_2D) {
+    std::vector<PartialShape> input_shapes{PartialShape{2, 12}, PartialShape{12}, PartialShape{12}};
+    const int64_t num_groups = 4;
+    element::Type elem_type = element::f32;
+
+    model = gen_model(input_shapes, elem_type, num_groups, 1e-3);
+    manager.register_pass<pass::GroupNormalizationDecomposition>();
+
+    // no decomposition
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/common/transformations/tests/op_conversions/group_normalization_decomposition_test.cpp:199]: Done

File: src/common/transformations/src/transformations/op_conversions/group_normalization_decomposition.cpp [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+//
+
+#include "transformations/op_conversions/group_normalization_decomposition.hpp"
+
+#include <memory>
+#include <vector>
+
+#include "itt.hpp"
+#include "openvino/core/rt_info.hpp"
+#include "openvino/opsets/opset12.hpp"
```
Comment (itikhono) [inline src/common/transformations/src/transformations/op_conversions/group_normalization_decomposition.cpp:12]: could you delete this include?

Comment (mitruska) [reply src/common/transformations/src/transformations/op_conversions/group_normalization_decomposition.cpp:12]: Is it preferred to include every op separately?

Comment (itikhono) [reply src/common/transformations/src/transformations/op_conversions/group_normalization_decomposition.cpp:12]: yes, this include adds ~5mb to each obj file size, the sum of all obj file sizes exceeds 4 gb limit on Windows in Debug configuration, so .lib file fails to create.

Comment (mitruska) [reply src/common/transformations/src/transformations/op_conversions/group_normalization_decomposition.cpp]: Ok, done

File: src/common/transformations/src/transformations/utils/utils.cpp (hunk 6 above)
Comment (praasz) [reply src/common/transformations/src/transformations/utils/utils.cpp:347]: The purpose of the ov::cmp::lt is safe compare between different types.

**Review by praasz — APPROVED** [review]

**Review by itikhono — APPROVED** [review]
